File: apps/syn-cli/src/syn_cli/commands/costs.py
"""Cost tracking commands — summary, sessions, executions."""

from __future__ import annotations

import typer
from rich.panel import Panel
from rich.table import Table

from syn_cli._output import (
    console,
    format_breakdown,
    format_cost,
    format_duration,
    format_timestamp,
    format_tokens,
    print_error,
)
from syn_cli.client import get_client
from syn_cli.commands._cost_models import (
    CostSummaryResponse,
    ExecutionCostResponse,
    SessionCostResponse,
)

app = typer.Typer(
    name="costs",
    help="View cost tracking data for sessions and executions",
    no_args_is_help=True,
)


def _safe_format_cost(value: str) -> str:
    """Format a cost string that may or may not already be $-prefixed."""
    if value.startswith("$"):
        return value
    try:
        return format_cost(value)
    except Exception:
        return value


def _handle_connect_error() -> None:
    from syn_cli.client import get_api_url

    print_error(f"Could not connect to API at {get_api_url()}")
    console.print("[dim]Make sure the API server is running.[/dim]")
    raise typer.Exit(1)


@app.command("summary")
def cost_summary() -> None:
    """Show aggregated cost summary across all sessions and executions."""
    try:
        with get_client() as client:
            resp = client.get("/costs/summary")
    except Exception:
        _handle_connect_error()
        return

    if resp.status_code != 200:
        print_error(resp.json().get("detail", f"HTTP {resp.status_code}"))
        raise typer.Exit(1)

    data = CostSummaryResponse(**resp.json())
    panel_text = (
        f"[bold]Total Cost:[/bold] {format_cost(data.total_cost_usd)}\n"
        f"[bold]Sessions:[/bold] {data.total_sessions}\n"
        f"[bold]Executions:[/bold] {data.total_executions}\n"
        f"[bold]Tokens:[/bold] {format_tokens(data.total_tokens)}\n"
        f"[bold]Tool Calls:[/bold] {data.total_tool_calls}"
    )
    console.print(Panel(panel_text, title="[cyan]Cost Summary[/cyan]", border_style="cyan"))

    if data.top_models:
        table = Table(title="Top Models")
        table.add_column("Model", style="cyan")
        table.add_column("Cost", justify="right")
        for entry in data.top_models:
            table.add_row(entry.get("model", "unknown"), entry.get("cost", "$0"))
        console.print(table)

    if data.top_sessions:
        table = Table(title="Top Sessions")
        table.add_column("Session", style="dim")
        table.add_column("Cost", justify="right")
        for entry in data.top_sessions:
            sid = entry.get("session_id", "unknown")
            table.add_row(sid[:12] + "..." if len(sid) > 12 else sid, entry.get("cost", "$0"))
        console.print(table)


@app.command("sessions")
def list_session_costs(
    execution_id: str | None = typer.Option(
        None, "--execution", "-e", help="Filter by execution ID"
    ),
    limit: int = typer.Option(50, "--limit", "-n", help="Max results", min=1, max=200),
) -> None:
    """List cost data for sessions."""
    try:
        with get_client() as client:
            params: dict[str, str | int] = {"limit": limit}
            if execution_id:
                params["execution_id"] = execution_id
            resp = client.get("/costs/sessions", params=params)
    except Exception:
        _handle_connect_error()
        return

    if resp.status_code != 200:
        print_error(resp.json().get("detail", f"HTTP {resp.status_code}"))
        raise typer.Exit(1)

    items = [SessionCostResponse(**s) for s in resp.json()]
    if not items:
        console.print("[dim]No session cost data found.[/dim]")
        return

    table = Table(title="Session Costs")
    table.add_column("Session ID", style="dim")
    table.add_column("Cost", justify="right")
    table.add_column("Tokens", justify="right")
    table.add_column("Duration", justify="right")
    table.add_column("Tools", justify="right")

    for s in items:
        table.add_row(
            s.session_id[:12] + "..." if len(s.session_id) > 12 else s.session_id,
            format_cost(s.total_cost_usd),
            format_tokens(s.total_tokens),
            format_duration(s.duration_ms),
            str(s.tool_calls),
        )
    console.print(table)


@app.command("session")
def show_session_cost(
    session_id: str = typer.Argument(..., help="Session ID"),
) -> None:
    """Show detailed cost breakdown for a session."""
    try:
        with get_client() as client:
            resp = client.get(f"/costs/sessions/{session_id}")
    except Exception:
        _handle_connect_error()
        return

    if resp.status_code != 200:
        print_error(resp.json().get("detail", f"HTTP {resp.status_code}"))
        raise typer.Exit(1)

    s = SessionCostResponse(**resp.json())
    panel_text = (
        f"[bold]Session:[/bold] {s.session_id}\n"
        f"[bold]Cost:[/bold] {format_cost(s.total_cost_usd)}\n"
        f"[bold]Tokens:[/bold] {format_tokens(s.total_tokens)} "
        f"(in: {format_tokens(s.input_tokens)}, out: {format_tokens(s.output_tokens)})\n"
        f"[bold]Tool Calls:[/bold] {s.tool_calls}  [bold]Turns:[/bold] {s.turns}\n"
        f"[bold]Duration:[/bold] {format_duration(s.duration_ms)}\n"
        f"[bold]Started:[/bold] {format_timestamp(s.started_at)}"
    )
    console.print(Panel(panel_text, title="[cyan]Session Cost Detail[/cyan]", border_style="cyan"))

    if s.cost_by_model:
        console.print(format_breakdown(s.cost_by_model, "Cost by Model", _safe_format_cost))
    if s.cost_by_tool:
        console.print(format_breakdown(s.cost_by_tool, "Cost by Tool", _safe_format_cost))


@app.command("executions")
def list_execution_costs(
    limit: int = typer.Option(50, "--limit", "-n", help="Max results", min=1, max=200),
) -> None:
    """List cost data for workflow executions."""
    try:
        with get_client() as client:
            resp = client.get("/costs/executions", params={"limit": limit})
    except Exception:
        _handle_connect_error()
        return

    if resp.status_code != 200:
        print_error(resp.json().get("detail", f"HTTP {resp.status_code}"))
        raise typer.Exit(1)

    items = [ExecutionCostResponse(**e) for e in resp.json()]
    if not items:
        console.print("[dim]No execution cost data found.[/dim]")
        return

    table = Table(title="Execution Costs")
    table.add_column("Execution ID", style="dim")
    table.add_column("Cost", justify="right")
    table.add_column("Sessions", justify="right")
    table.add_column("Tokens", justify="right")

    for e in items:
        table.add_row(
            e.execution_id[:12] + "..." if len(e.execution_id) > 12 else e.execution_id,
            format_cost(e.total_cost_usd),
            str(e.session_count),
            format_tokens(e.total_tokens),
        )
    console.print(table)


@app.command("execution")
def show_execution_cost(
    execution_id: str = typer.Argument(..., help="Execution ID"),
) -> None:
    """Show detailed cost breakdown for an execution."""
    try:
        with get_client() as client:
            resp = client.get(f"/costs/executions/{execution_id}")
    except Exception:
        _handle_connect_error()
        return

    if resp.status_code != 200:
        print_error(resp.json().get("detail", f"HTTP {resp.status_code}"))
        raise typer.Exit(1)

    e = ExecutionCostResponse(**resp.json())
    panel_text = (
        f"[bold]Execution:[/bold] {e.execution_id}\n"
        f"[bold]Cost:[/bold] {format_cost(e.total_cost_usd)}\n"
        f"[bold]Sessions:[/bold] {e.session_count}\n"
        f"[bold]Tokens:[/bold] {format_tokens(e.total_tokens)} "
        f"(in: {format_tokens(e.input_tokens)}, out: {format_tokens(e.output_tokens)})\n"
        f"[bold]Duration:[/bold] {format_duration(e.duration_ms)}\n"
        f"[bold]Started:[/bold] {format_timestamp(e.started_at)}"
    )
    console.print(
        Panel(panel_text, title="[cyan]Execution Cost Detail[/cyan]", border_style="cyan")
    )

    if e.cost_by_phase:
        console.print(format_breakdown(e.cost_by_phase, "Cost by Phase", _safe_format_cost))
    if e.cost_by_model:
        console.print(format_breakdown(e.cost_by_model, "Cost by Model", _safe_format_cost))
    if e.cost_by_tool:
        console.print(format_breakdown(e.cost_by_tool, "Cost by Tool", _safe_format_cost))

File: apps/syn-cli-node/src/commands/costs.ts
/**
 * Cost tracking commands — summary, sessions, session, executions, execution.
 * Port of apps/syn-cli/src/syn_cli/commands/costs.py
 */

import { CommandGroup, type CommandDef, type ParsedArgs } from "../framework/command.js";
import { CLIError } from "../framework/errors.js";
import { api, unwrap } from "../client/typed.js";
import { print, printError, printDim } from "../output/console.js";
import { style, BOLD, CYAN, DIM } from "../output/ansi.js";
import { formatCost, formatDuration, formatTimestamp, formatTokens, formatBreakdown } from "../output/format.js";
import { Table } from "../output/table.js";


function safeCost(v: string): string {
  if (v.startsWith("$")) return v;
  try { return formatCost(v); } catch { return v; }
}

const summaryCommand: CommandDef = {
  name: "summary",
  description: "Show aggregated cost summary",
  handler: async () => {
    const d = unwrap(await api.GET("/costs/summary"), "Get cost summary");

    print(style("Cost Summary", CYAN));
    print(`  ${style("Total Cost:", BOLD)} ${formatCost(d.total_cost_usd)}`);
    print(`  ${style("Sessions:", BOLD)} ${d.total_sessions}`);
    print(`  ${style("Executions:", BOLD)} ${d.total_executions}`);
    print(`  ${style("Tokens:", BOLD)} ${formatTokens(d.total_tokens)}`);
    print(`  ${style("Tool Calls:", BOLD)} ${d.total_tool_calls}`);

    const topModels = (d.top_models ?? []) as Record<string, string>[];
    if (topModels.length > 0) {
      const table = new Table({ title: "Top Models" });
      table.addColumn("Model", { style: CYAN });
      table.addColumn("Cost", { align: "right" });
      for (const e of topModels) table.addRow(e["model"] ?? "unknown", e["cost"] ?? "$0");
      table.print();
    }

    const topSessions = (d.top_sessions ?? []) as Record<string, string>[];
    if (topSessions.length > 0) {
      const table = new Table({ title: "Top Sessions" });
      table.addColumn("Session", { style: DIM });
      table.addColumn("Cost", { align: "right" });
      for (const e of topSessions) {
        const sid = e["session_id"] ?? "unknown";
        table.addRow(sid.length > 12 ? sid.slice(0, 12) + "..." : sid, e["cost"] ?? "$0");
      }
      table.print();
    }
  },
};

const sessionsCommand: CommandDef = {
  name: "sessions",
  description: "List cost data for sessions",
  options: {
    execution: { type: "string", short: "e", description: "Filter by execution ID" },
    limit: { type: "string", short: "n", description: "Max results", default: "50" },
  },
  handler: async (parsed: ParsedArgs) => {
    const limit = Number((parsed.values["limit"] as string | undefined) ?? "50");
    const items = unwrap(await api.GET("/costs/sessions", {
      params: {
        query: {
          execution_id: (parsed.values["execution"] as string | undefined) ?? null,
          limit,
        },
      },
    }), "List session costs");

    if (items.length === 0) { printDim("No session cost data found."); return; }

    const table = new Table({ title: "Session Costs" });
    table.addColumn("Session ID", { style: DIM });
    table.addColumn("Cost", { align: "right" });
    table.addColumn("Tokens", { align: "right" });
    table.addColumn("Duration", { align: "right" });
    table.addColumn("Tools", { align: "right" });

    for (const s of items) {
      const sid = s.session_id;
      table.addRow(
        sid.length > 12 ? sid.slice(0, 12) + "..." : sid,
        formatCost(s.total_cost_usd),
        formatTokens(s.total_tokens),
        formatDuration(s.duration_ms),
        String(s.tool_calls),
      );
    }
    table.print();
  },
};

const sessionDetailCommand: CommandDef = {
  name: "session",
  description: "Show detailed cost breakdown for a session",
  args: [{ name: "session-id", description: "Session ID", required: true }],
  handler: async (parsed: ParsedArgs) => {
    const id = parsed.positionals[0];
    if (!id) { printError("Missing session-id"); throw new CLIError("Missing argument", 1); }

    const s = unwrap(await api.GET("/costs/sessions/{session_id}", { params: { path: { session_id: id } } }), "Get session cost");

    print(style("Session Cost Detail", CYAN));
    print(`  ${style("Session:", BOLD)} ${s.session_id}`);
    print(`  ${style("Cost:", BOLD)} ${formatCost(s.total_cost_usd)}`);
    print(`  ${style("Tokens:", BOLD)} ${formatTokens(s.total_tokens)} (in: ${formatTokens(s.input_tokens)}, out: ${formatTokens(s.output_tokens)})`);
    print(`  ${style("Tool Calls:", BOLD)} ${s.tool_calls}  ${style("Turns:", BOLD)} ${s.turns}`);
    print(`  ${style("Duration:", BOLD)} ${formatDuration(s.duration_ms)}`);
    print(`  ${style("Started:", BOLD)} ${formatTimestamp(s.started_at)}`);

    if (s.cost_by_model && Object.keys(s.cost_by_model).length > 0) print(formatBreakdown(s.cost_by_model, "Cost by Model", safeCost));
    if (s.cost_by_tool && Object.keys(s.cost_by_tool).length > 0) print(formatBreakdown(s.cost_by_tool, "Cost by Tool", safeCost));
  },
};

const executionsCommand: CommandDef = {
  name: "executions",
  description: "List cost data for workflow executions",
  options: { limit: { type: "string", short: "n", description: "Max results", default: "50" } },
  handler: async (parsed: ParsedArgs) => {
    const limit = Number((parsed.values["limit"] as string | undefined) ?? "50");
    const items = unwrap(await api.GET("/costs/executions", { params: { query: { limit } } }), "List execution costs");

    if (items.length === 0) { printDim("No execution cost data found."); return; }

    const table = new Table({ title: "Execution Costs" });
    table.addColumn("Execution ID", { style: DIM });
    table.addColumn("Cost", { align: "right" });
    table.addColumn("Sessions", { align: "right" });
    table.addColumn("Tokens", { align: "right" });

    for (const e of items) {
      const eid = e.execution_id;
      table.addRow(
        eid.length > 12 ? eid.slice(0, 12) + "..." : eid,
        formatCost(e.total_cost_usd),
        String(e.session_count),
        formatTokens(e.total_tokens),
      );
    }
    table.print();
  },
};

const executionDetailCommand: CommandDef = {
  name: "execution",
  description: "Show detailed cost breakdown for an execution",
  args: [{ name: "execution-id", description: "Execution ID", required: true }],
  handler: async (parsed: ParsedArgs) => {
    const id = parsed.positionals[0];
    if (!id) { printError("Missing execution-id"); throw new CLIError("Missing argument", 1); }

    const e = unwrap(await api.GET("/costs/executions/{execution_id}", { params: { path: { execution_id: id } } }), "Get execution cost");

    print(style("Execution Cost Detail", CYAN));
    print(`  ${style("Execution:", BOLD)} ${e.execution_id}`);
    print(`  ${style("Cost:", BOLD)} ${formatCost(e.total_cost_usd)}`);
    print(`  ${style("Sessions:", BOLD)} ${e.session_count}`);
    print(`  ${style("Tokens:", BOLD)} ${formatTokens(e.total_tokens)} (in: ${formatTokens(e.input_tokens)}, out: ${formatTokens(e.output_tokens)})`);
    print(`  ${style("Duration:", BOLD)} ${formatDuration(e.duration_ms)}`);
    print(`  ${style("Started:", BOLD)} ${formatTimestamp(e.started_at)}`);

    if (e.cost_by_phase && Object.keys(e.cost_by_phase).length > 0) print(formatBreakdown(e.cost_by_phase, "Cost by Phase", safeCost));
    if (e.cost_by_model && Object.keys(e.cost_by_model).length > 0) print(formatBreakdown(e.cost_by_model, "Cost by Model", safeCost));
    if (e.cost_by_tool && Object.keys(e.cost_by_tool).length > 0) print(formatBreakdown(e.cost_by_tool, "Cost by Tool", safeCost));
  },
};

export const costsGroup = new CommandGroup("costs", "View cost tracking data for sessions and executions");
costsGroup.command(summaryCommand).command(sessionsCommand).command(sessionDetailCommand).command(executionsCommand).command(executionDetailCommand);
